File: src/db/repository.py
from datetime import datetime

import structlog
from sqlalchemy import create_engine, select
from sqlalchemy.orm import Session

from src.db.models import Base, SendLog, SentArticle

log = structlog.get_logger()


class ArticleRepository:
    def __init__(self, database_url: str) -> None:
        self._engine = create_engine(database_url)
        Base.metadata.create_all(self._engine)

    def is_sent(self, url: str) -> bool:
        with Session(self._engine) as session:
            result = session.execute(
                select(SentArticle.id).where(SentArticle.url == url).limit(1)
            ).first()
            return result is not None

    def filter_unsent(self, urls: list[str]) -> list[str]:
        if not urls:
            return []
        with Session(self._engine) as session:
            sent = set(
                row[0]
                for row in session.execute(
                    select(SentArticle.url).where(SentArticle.url.in_(urls))
                ).all()
            )
        return [u for u in urls if u not in sent]

    def mark_sent(self, url: str, title: str, source: str) -> None:
        with Session(self._engine) as session:
            existing = session.execute(
                select(SentArticle).where(SentArticle.url == url)
            ).scalar_one_or_none()
            if existing is None:
                session.add(SentArticle(url=url, title=title, source=source))
                session.commit()

    def record_log(
        self,
        run_at: datetime,
        article_count: int,
        recipient_count: int,
        status: str,
        error_message: str | None = None,
    ) -> None:
        with Session(self._engine) as session:
            session.add(
                SendLog(
                    run_at=run_at,
                    article_count=article_count,
                    recipient_count=recipient_count,
                    status=status,
                    error_message=error_message,
                )
            )
            session.commit()

File: src/db/repository.ts
import fs from "fs";
import path from "path";

export interface SentArticle {
  url: string;
  title: string;
  source: string;
  sentAt: string;
}

export interface SendLog {
  runAt: string;
  articleCount: number;
  recipientCount: number;
  status: "success" | "partial" | "failed";
  errorMessage?: string;
  createdAt: string;
}

interface StoreData {
  sentArticles: SentArticle[];
  sendLogs: SendLog[];
}

export class ArticleRepository {
  private dataPath: string;
  private data: StoreData;

  constructor(dataPath: string) {
    this.dataPath = dataPath;
    this.data = this.load();
  }

  private load(): StoreData {
    const dir = path.dirname(this.dataPath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    if (!fs.existsSync(this.dataPath)) {
      return { sentArticles: [], sendLogs: [] };
    }
    try {
      return JSON.parse(fs.readFileSync(this.dataPath, "utf-8")) as StoreData;
    } catch {
      return { sentArticles: [], sendLogs: [] };
    }
  }

  private save(): void {
    fs.writeFileSync(this.dataPath, JSON.stringify(this.data, null, 2), "utf-8");
  }

  filterUnsent(urls: string[]): string[] {
    const sentSet = new Set(this.data.sentArticles.map((a) => a.url));
    return urls.filter((url) => !sentSet.has(url));
  }

  markSent(articles: Array<{ url: string; title: string; source: string }>): void {
    const now = new Date().toISOString();
    for (const article of articles) {
      this.data.sentArticles.push({ ...article, sentAt: now });
    }
    this.save();
  }

  addLog(log: Omit<SendLog, "createdAt">): void {
    this.data.sendLogs.push({ ...log, createdAt: new Date().toISOString() });
    this.save();
  }
}
